src/packet/header/fragment.rs

```diff
@@ -0,0 +1,128 @@
+use super::{HeaderParser, HeaderReader};
+use super::PacketHeader;
+use error::{Result, NetworkError};
+use net::constants::{FRAGMENT_HEADER_SIZE};
+
+use std::io::{self, Cursor, Error, ErrorKind, Write};
+use byteorder::{BigEndian, ReadBytesExt, WriteBytesExt};
+
+#[derive(Copy, Clone, Debug)]
+/// This header represents an fragmented packet header.
+pub struct FragmentHeader {
+    pub sequence: u16,
+    pub id: u8,
+    pub num_fragments: u8,
+    pub packet_header: Option<PacketHeader>,
+}
+
+impl FragmentHeader {
+    /// Create new fragment with the given packet header
+    pub fn new(id: u8, num_fragments: u8, packet_header: PacketHeader) -> Self {
+        FragmentHeader { id, num_fragments, packet_header: Some(packet_header), sequence: packet_header.seq }
+    }
+
+    /// Get the size of this header.
+    pub fn size(&self) -> u8
+    {
+        if self.id == 0 {
+            match self.packet_header
+            {
+                Some(header) => header.size() + FRAGMENT_HEADER_SIZE,
+                None => {
+                    error!("Attempting to retrieve size on a 0 ID packet with no packet header");
+                    0
+                }
+            }
+        } else {
+            FRAGMENT_HEADER_SIZE
+        }
+    }
+}
+
+impl HeaderParser for FragmentHeader
+{
+    type Output = io::Result<Vec<u8>>;
+
+    fn parse(&self) -> <Self as HeaderParser>::Output {
+
+        let mut wtr = Vec::new();
+        wtr.write_u8(1)?;
+        wtr.write_u16::<BigEndian>(self.sequence)?;
+        wtr.write_u8(self.id)?;
+        wtr.write_u8(self.num_fragments)?;
+
+        if self.id == 0 {
+            match self.packet_header
+            {
+                Some(header) => {
+                    wtr.write(&header.parse()?)?;
+                },
+                None => {
+                    return Err(Error::new(ErrorKind::Other, "Invalid fragment header"));
```

Comment:
Should have an error type for this

Reply:
There is one problem of doing that you have to see the context of the whole function. 

Every 'write' in that function returns an `io::Result`. So we could paste this line `.map_err(|| NetworkError::InvalidPacketHeader.into()?);` behind each write function or we could simply return `io::Result` and let the caller of that function maps the error to `.map_err(|| NetworkError::InvalidPacketHeader.into()?);` 

I think that is better otherwise this function will look horrible.

Reply:
In reality we should have a Error::from implemented for `io::Result`